src/hotspot/os_cpu/linux_aarch64/vm_version_linux_aarch64.cpp

```diff
@@ -168,24 +168,34 @@ void VM_Version::get_os_cpu_info() {
   }
 }
 
-void VM_Version::get_compatible_board(char *buf, int buflen) {
+static bool read_fully(const char *fname, char *buf, int buflen) {
   assert(buf != NULL, "invalid argument");
   assert(buflen >= 1, "invalid argument");
-  *buf = '\0';
-  int fd = open("/proc/device-tree/compatible", O_RDONLY);
+  int fd = open(fname, O_RDONLY);
   if (fd != -1) {
     ssize_t read_sz = read(fd, buf, buflen - 1);
-    if (read_sz > 0) {
+    close(fd);
+    if ((read_sz > 0) && isgraph(*buf)) {
```

Comment:
What is `isgraph(*buf)` doing here? Is the problem that the file might contain binary data?

Reply:
In case of AWS A1 2x.large, board_name has `\n` (0x0A) only. This code intent to ignore it.

Reply:
Please be more explicit. This is (perhaps obviously?) a workaround for a bug, And t.his code is itself a bug, because it doesn't check for the condition you care about. So, please check explicitly for this condition: `\n` only, and nothing else. And leave a comment explaining why.

Reply:
One other thing I missed last time around: why is there a permanently-allocated 4k buffer for this code that it only used for jfr?

Reply:
I pushed new commit with some refactoring.

> Please be more explicit. This is (perhaps obviously?) a workaround for a bug, And t.his code is itself a bug, because it doesn't check for the condition you care about. So, please check explicitly for this condition: \n only, and nothing else. And leave a comment explaining why.

I've done so in new commit.

> One other thing I missed last time around: why is there a permanently-allocated 4k buffer for this code that it only used for jfr?

Some features (e.g. hs_err log, VM.info dcmd) dumps /proc/cpuinfo, so I guess this feature is used JFR only now. However CPU description has different information (for AArch64 at least), so we might be able to use it in other place, but it is out of scope this PR.

Reply:
Looks better. I guess the problem here is that we don't know how robust we need to make this against future bugs.
Maybe we should just concatenate all three files?

Reply:
> Maybe we should just concatenate all three files?

Do you mean we should concatenate "compatible", "board_name", and "product_name"? It's impossible. We can see "compatible" file on the device which uses device tree, however we cannot see both "board_name" and "product_name" on it. Thus I prefer to attempt to read them sequentially.

Reply:
> > Maybe we should just concatenate all three files?
> 
> Do you mean we should concatenate "compatible", "board_name", and "product_name"? It's impossible.

Of course it isn't impossible. But if you really want to skip a file which is just "\n", then please check for the file being just "\n", not a file with "\n" as its first character.

Reply:
I pushed new commit to check the contents whether it is just "\n". Could you review again?

Reply:
@theRealAph The latest change uses `strncmp(buf, "\n", 2)` to check whether the content just has "\n". Are you ok this change?